src/cairo1-compiler.ts

```diff
@@ -14,48 +25,110 @@ export const exec = (args: string) => {
     } as ProcessResult;
 };
 
-export class DockerCairo1Compiler {
-    image: Image;
-    sources: string[];
-    compilerArgs: string[];
+export class CairoCompilerDownloader {
+    private isInitialized: Promise<void>;
+    compilerDownloadPath: string;
+    compilerVersion: string;
 
-    constructor(image: Image, sources: string[], cairo1CompilerArgs?: string[]) {
-        this.image = image;
-        this.sources = sources;
-        this.compilerArgs = cairo1CompilerArgs;
+    constructor(rootPath: string, starknet: StarknetConfig) {
+        this.isInitialized = this.initialize(rootPath, starknet);
     }
 
-    protected getDockerArgs(): BindsMap {
-        const binds: BindsMap = {};
-        for (const source of this.sources) {
-            binds[source] = source;
+    async initialize(rootPath: string, starknet: StarknetConfig) {
+        this.compilerDownloadPath = starknet?.cairo1BinDir || path.join(rootPath, "cairo-compiler");
+        this.compilerVersion = starknet?.compilerVersion || (await this.getCompilerVersion());
+    }
+
+    async handleCompilerDownload(): Promise<void> {
+        await this.isInitialized;
+        if (fs.existsSync(this.getBinDirPath())) {
```

Comment:
Here's the check I think we should be doing:
- get the binary path
- check if the binary file exists
- check if the binary file is executable by running the `starknet-compile --version` command, but simply relying on the exit code of the process being 0 (no str matching)
  - doing `output.contains(str)` can be wrong (e.g. `1.1.1` is a subset of `1.1.10`)

Reply:
So, I avoided string matching and now it checks if the version is strictly equal to what is specified by the user. We need this because there could be an older/different version of the compiler installed already but different one in the `hardhat.config.ts` file.

https://github.com/0xSpaceShard/starknet-hardhat-plugin/pull/402/commits/5efc50a7c425e58b74d93a39d4609ca0e4ecdc76